Signal/src/ViewControllers/CallViewController.swift

```diff
@@ -961,6 +962,29 @@ class CallViewController: OWSViewController, CallObserver, CallServiceObserver {
         self.updateCallUI(callState: call.state)
     }
 
+    // MARK: CallAudioServiceDelegate
+
+    func callAudioService(_ callAudioService: CallAudioService, didUpdateIsSpeakerphoneEnabled isSpeakerphoneEnabled: Bool) {
+        AssertIsOnMainThread()
+
+        updateAudioSourceButtonIsSelected()
+    }
+
+    func callAudioServiceDidChangeAudioSession(_ callAudioService: CallAudioService) {
+        AssertIsOnMainThread()
+
+        // Which sources are available depends on the state of your Session.
+        // When the audio session is not yet in PlayAndRecord none are available
+        // Then if we're in speakerphone, bluetooth isn't available.
+        // So we accrue all possible audio sources in a set, and that list lives as longs as the CallViewController
+        // The downside of this is that if you e.g. unpair your bluetooth mid call, it will still appear as an option
+        // until your next call.
+        // FIXME: There's got to be a better way, but this is where I landed after a bit of work, and seems to work
+        // pretty well in practice.
+        let availableInputs = callAudioService.availableInputs
+        self.allAudioSources.formUnion(availableInputs)
```

Comment:
Seems like we'll need to test this on iOS 8, 9, 10, 11.  Please make sure to specify that in the test plan.

Reply:
Sure thing.